File: C4_InnerTube/InnerTubeVB/SharedUtilities/SettingService.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.IO

Namespace SharedUtilities
	Public NotInheritable Class SettingService

        Private Shared privateVideoPath As String
		Public Shared Property VideoPath() As String
			Get
				Return privateVideoPath
			End Get
			Set(ByVal value As String)
				privateVideoPath = value
			End Set
		End Property
		Public Shared SubPath As String

        Private Shared privateSettingPath As String
		Public Shared Property SettingPath() As String
			Get
				Return privateSettingPath
			End Get
			Set(ByVal value As String)
				privateSettingPath = value
			End Set
		End Property

		Public Shared Function Load() As Setting
			SetPath()

			If File.Exists(SettingPath) Then
				Dim serial = New Serializer(Of Setting)()
				Return serial.Deserialize(SettingPath)
			Else
				Return BuildDefaultSettings()
			End If
		End Function

		Public Shared Sub Save(ByVal changes As Setting)
			SetPath()
			Dim serial = New Serializer(Of Setting)()
			serial.Serialize(changes, SettingPath)
		End Sub

		Private Shared Sub SetPath()
			If String.IsNullOrEmpty(VideoPath) Then
				VideoPath = FileHelper.BuildPath(SubPath)
			End If

			If String.IsNullOrEmpty(SubPath) Then
				SubPath = Path.Combine(SubPath, FileHelper.DefaultSettingsName)
			End If

			SettingPath = Path.Combine(SubPath, FileHelper.DefaultSettingsName)
		End Sub

		Private Shared Function BuildDefaultSettings() As Setting
			Dim s As New Setting()
			s.AppName = FileHelper.DefaultAppName
			s.iTunesInstalled = False
			s.ZuneInstalled = False
			s.FirstRun = True
			s.SubPath = SubPath
			s.InnerTubeFeedFile = Path.Combine(SubPath, FileHelper.BuildXmlName())
			s.VideoPath = VideoPath
			s.UpdateFeedPoolThreads = 3
            s.DownloadPoolThreads = 5
			s.ConversionPoolThreads = 2

			Return s
		End Function

	End Class
End Namespace

File: C4_InnerTube/InnerTubeCS/SharedUtilities/SettingService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.IO;

namespace SharedUtilities
{
    public static class SettingService
    {

        public static string VideoPath { get; set; }
        public static string SubPath;
        public static string SettingPath { get; set; }


        public static Setting Load()
        {
            SetPath();

            if (File.Exists(SettingPath))
            {
                var serial = new Serializer<Setting>();
                return serial.Deserialize(SettingPath);
            }
            else
            {
                return BuildDefaultSettings();
            }
        }

        public static void Save(Setting changes)
        {
            SetPath();
            var serial = new Serializer<Setting>();
            serial.Serialize(changes, SettingPath);
        }

        private static void SetPath()
        {
            if (String.IsNullOrEmpty(VideoPath))
            {
                VideoPath = FileHelper.BuildPath(ref SubPath);
            }

            if (String.IsNullOrEmpty(SubPath))
            {
                SubPath = Path.Combine(SubPath, FileHelper.DefaultSettingsName);
            }

            SettingPath = Path.Combine(SubPath, FileHelper.DefaultSettingsName);
        }

        private static Setting BuildDefaultSettings()
        {
            Setting s = new Setting();
            s.AppName = FileHelper.DefaultAppName;
            s.iTunesInstalled = false;
            s.ZuneInstalled = false;
            s.FirstRun = true;
            s.SubPath = SubPath;
            s.InnerTubeFeedFile = Path.Combine(SubPath, FileHelper.BuildXmlName());
            s.VideoPath = VideoPath;
            s.UpdateFeedPoolThreads = 3;
            s.DownloadPoolThreads = 4;
            s.ConversionPoolThreads = 2;

            return s;
        }

    }
}
